### PR TITLE
storage: disallow splits in meta2 ranges

### PR DESCRIPTION
**Pull request by petermattis**
Truncate range lookup scans to the range.

See #16266

### CODE REVIEW

**Comment by petermattis** [comment]
This is a WIP. The test I wrote fails in the way I expect it to fail. Now to figure out what to do about it.

**Comment by petermattis** [comment]
I need to expand the tests somewhat. In particular, I want to test sized-based splitting of meta ranges and verify that selected split point is as expected.

**Comment by bdarnell** [comment]
Reviewed 3 of 4 files at r1.
Review status: 3 of 4 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/keys/keys.go, line 504 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17565#-KrBbxFFmmMbZx7iP0Hx:-KrBbxFFmmMbZx7iP0Hy:b-yb4qao) ([raw file](https://github.com/cockroachdb/cockroach/blob/90bcd3f2a928b2a71ab4bc81c6f135192fbbce94/pkg/keys/keys.go#L504)):*
> ```Go
> // valid RangeMetaKey as defined by validateRangeMetaKey.
> // TODO(tschottdorf): a lot of casting going on inside.
> func MetaScanBounds(key roachpb.RKey, span roachpb.RSpan) (roachpb.Key, roachpb.Key, error) {
> ```

Document the `span` parameter. 

---

*[pkg/storage/replica_command.go, line 1173 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17565#-KrBeuN2aulInY3CwpWA:-KrBeuN2aulInY3CwpWB:bst19ch) ([raw file](https://github.com/cockroachdb/cockroach/blob/90bcd3f2a928b2a71ab4bc81c6f135192fbbce94/pkg/storage/replica_command.go#L1173)):*
> ```Go
> 		// args.Key. Scan for both the requested key and the keys immediately
> 		// afterwards, up to MaxRanges.
> 		startKey, endKey, err := keys.MetaScanBounds(key, desc.RSpan())
> ```

Maybe MetaScanBounds should return an RSpan and then we can use `RSpan.Intersect(desc)` to constrain it (or, if what MetaScanBounds is doing is different from that, we need to document it)

---

*[pkg/storage/replica_command.go, line 1269 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17565#-KrBfBppqg6UB0Xghbsg:-KrBfBppqg6UB0Xghbsh:bdqai78) ([raw file](https://github.com/cockroachdb/cockroach/blob/90bcd3f2a928b2a71ab4bc81c6f135192fbbce94/pkg/storage/replica_command.go#L1269)):*
> ```Go
> 	for _, kv := range kvs {
> 		if !desc.RSpan().ContainsKey(roachpb.RKey(kv.Key)) {
> 			// Skip keys which lie outside of our range. This can happen when Reverse
> ```

How? Isn't it constrained in both cases?

---

*[pkg/storage/replica_command.go, line 2611 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17565#-KrBgRIHE4_q3IIYyifd:-KrBgRIHE4_q3IIYyife:b31o8qt) ([raw file](https://github.com/cockroachdb/cockroach/blob/90bcd3f2a928b2a71ab4bc81c6f135192fbbce94/pkg/storage/replica_command.go#L2611)):*
> ```Go
> 	var splitKey roachpb.RKey
> 	{
> 		if len(args.SplitKey) > 0 {
> ```

Move this up a couple of lines - it doesn't need to be inside this variable scope block. 

---

*[pkg/storage/replica_command.go, line 2622 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17565#-KrBgbsPGryxqvm4MzDi:-KrBgbsPGryxqvm4MzDj:b-szz3xz) ([raw file](https://github.com/cockroachdb/cockroach/blob/90bcd3f2a928b2a71ab4bc81c6f135192fbbce94/pkg/storage/replica_command.go#L2622)):*
> ```Go
> 
> 		metaSpan := roachpb.RSpan{Key: roachpb.RKeyMin, EndKey: roachpb.RKey(keys.SystemPrefix)}
> 		isMetaRange := metaSpan.ContainsKeyRange(desc.StartKey, desc.EndKey)
> ```

Use `keys.MetaMax` instead of `keys.SystemPrefix`. This could just be `if desc.StartKey < keys.MetaMax`. 

---

*[pkg/storage/replica_command.go, line 2630 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17565#-KrBhKky11QVCWM97IYw:-KrBhKky11QVCWM97IYx:b-mqsgoo) ([raw file](https://github.com/cockroachdb/cockroach/blob/90bcd3f2a928b2a71ab4bc81c6f135192fbbce94/pkg/storage/replica_command.go#L2630)):*
> ```Go
> 
> 			if len(args.SplitKey) > 0 {
> 				// We're splitting a meta range and the split key was specified. Meta
> ```

Transforming the requested split key has been problematic in the past. It's better to return an error here if an invalid split key is requested and put the burden on the places where the split keys are chosen.

But what's an invalid split key? I don't think this is well-defined. To use the example from the commit message, `d\0` is *also* a valid key. If the range `[d, e)` splits again at `d\0` (unlikely, but possible if there are tons of mvcc versions at `d`), the bug returns because the meta record for `[d, d\0)`lives on the wrong side of the (existing) split. I don't think we can find "safe" split points this way.

I think we need to fix this on the DistSender side, by adjusting the addressing on the query and/or retrying when we encounter a split (and in the meantime, for 1.1 I'd just disallow meta2 splits). 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17565)*

**Comment by petermattis** [comment]
Review status: 3 of 4 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica_command.go, line 2630 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17565#-KrBhKky11QVCWM97IYw:-KrBlwP2zTdvtOFRn4cO:bhlad73) ([raw file](https://github.com/cockroachdb/cockroach/blob/90bcd3f2a928b2a71ab4bc81c6f135192fbbce94/pkg/storage/replica_command.go#L2630)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Transforming the requested split key has been problematic in the past. It's better to return an error here if an invalid split key is requested and put the burden on the places where the split keys are chosen.

But what's an invalid split key? I don't think this is well-defined. To use the example from the commit message, `d\0` is *also* a valid key. If the range `[d, e)` splits again at `d\0` (unlikely, but possible if there are tons of mvcc versions at `d`), the bug returns because the meta record for `[d, d\0)`lives on the wrong side of the (existing) split. I don't think we can find "safe" split points this way.

I think we need to fix this on the DistSender side, by adjusting the addressing on the query and/or retrying when we encounter a split (and in the meantime, for 1.1 I'd just disallow meta2 splits). 
</blockquote></details>

Hmm, I think you've highlighted a more fundamental problem:

```
range 1 [a, d\0):
  b -> [a, b)
  c -> [b, c)
  d -> [c, d)
range 2 [d\0, g\0):
  e -> [d, e)
  f -> [e, f)
  g -> [f, g)
```

Lookup of the range containing `d` would send a request to range 1, not range 2. So the problem returns, though just for the key on the boundary between meta ranges.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17565)*

**Comment by petermattis** [comment]
Switched to disabling splits in meta2. PTAL.

---

Review status: 0 of 6 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[pkg/keys/keys.go, line 504 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17565#-KrBbxFFmmMbZx7iP0Hx:-KrCQROYzglRidKwDCvg:b2zh28t) ([raw file](https://github.com/cockroachdb/cockroach/blob/90bcd3f2a928b2a71ab4bc81c6f135192fbbce94/pkg/keys/keys.go#L504)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Document the `span` parameter. 
</blockquote></details>

Took the approach of returning an `RSpan` and using `RSpan.Intersect`.

---

*[pkg/storage/replica_command.go, line 1173 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17565#-KrBeuN2aulInY3CwpWA:-KrCQV17zhhWrGNjHRzP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/90bcd3f2a928b2a71ab4bc81c6f135192fbbce94/pkg/storage/replica_command.go#L1173)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Maybe MetaScanBounds should return an RSpan and then we can use `RSpan.Intersect(desc)` to constrain it (or, if what MetaScanBounds is doing is different from that, we need to document it)
</blockquote></details>

Done.

---

*[pkg/storage/replica_command.go, line 1269 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17565#-KrBfBppqg6UB0Xghbsg:-KrCMDkpjIu-aK7HFWBO:bxxopsa) ([raw file](https://github.com/cockroachdb/cockroach/blob/90bcd3f2a928b2a71ab4bc81c6f135192fbbce94/pkg/storage/replica_command.go#L1269)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

How? Isn't it constrained in both cases?
</blockquote></details>

I had a bug.

---

*[pkg/storage/replica_command.go, line 2611 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17565#-KrBgRIHE4_q3IIYyifd:-KrCQODyYlkHr2vAeg4Q:bp2dcqx) ([raw file](https://github.com/cockroachdb/cockroach/blob/90bcd3f2a928b2a71ab4bc81c6f135192fbbce94/pkg/storage/replica_command.go#L2611)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Move this up a couple of lines - it doesn't need to be inside this variable scope block. 
</blockquote></details>

Reverted this part of the change.

---

*[pkg/storage/replica_command.go, line 2622 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17565#-KrBgbsPGryxqvm4MzDi:-KrCQPgtZWkKnqXpnfPu:bp2dcqx) ([raw file](https://github.com/cockroachdb/cockroach/blob/90bcd3f2a928b2a71ab4bc81c6f135192fbbce94/pkg/storage/replica_command.go#L2622)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Use `keys.MetaMax` instead of `keys.SystemPrefix`. This could just be `if desc.StartKey < keys.MetaMax`. 
</blockquote></details>

Reverted this part of the change.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17565#-:-KrCQY9Ggox2nogVaSBu:b-6706s)*

**Comment by tamird** [comment]
Reviewed 7 of 7 files at r2.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks pending.

---

*[pkg/keys/keys_test.go, line 210 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17565#-KrCRBHysMpuibhe0na2:-KrCRBHysMpuibhe0na3:b-fhuimh) ([raw file](https://github.com/cockroachdb/cockroach/blob/e74b74a0ba712d106321ae5cedca800821a67b09/pkg/keys/keys_test.go#L210)):*
> ```Go
> 
> 	testCases := []struct {
> 		key, expStart, expEnd []byte
> ```

should these all be roachpb.Key instead of `[]byte`? ditto Reverse

---

*[pkg/keys/spans.go, line 23 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17565#-KrCRQtRQsYwKWMvm2Ye:-KrCRQtRQsYwKWMvm2Yf:b-p8p61v) ([raw file](https://github.com/cockroachdb/cockroach/blob/e74b74a0ba712d106321ae5cedca800821a67b09/pkg/keys/spans.go#L23)):*
> ```Go
> 	//
> 	// TODO(peter): Allow splitting of meta2 records. This requires enhancement
> 	// to range lookup:
> ```

was there supposed to be something after this colon?

---

*[pkg/kv/split_test.go, line 108 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17565#-KrCRU08y3Bg-FSB5mKP:-KrCRU08y3Bg-FSB5mKQ:bczeone) ([raw file](https://github.com/cockroachdb/cockroach/blob/e74b74a0ba712d106321ae5cedca800821a67b09/pkg/kv/split_test.go#L108)):*
> ```Go
> 	ctx := context.TODO()
> 
> 	// TODO(peter): revert when we allow meta2 splitting.
> ```

can you reference an issue? (throughout)

---

*[pkg/storage/replica_command.go, line 1353 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17565#-KrCRyhJrDcg55VRJJnN:-KrCRyhJrDcg55VRJJnO:b1zm3jc) ([raw file](https://github.com/cockroachdb/cockroach/blob/e74b74a0ba712d106321ae5cedca800821a67b09/pkg/storage/replica_command.go#L1353)):*
> ```Go
> 		// See #16266.
> 		var buf bytes.Buffer
> 		fmt.Fprintf(&buf, "range lookup of meta key '%s' [%x] found only non-matching ranges:",
> ```

nit: `'%[1]s' [%[1]x]` so you pass args.Key just once.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17565)*

**Comment by petermattis** [comment]
Review status: 2 of 6 files reviewed at latest revision, 10 unresolved discussions.

---

*[pkg/keys/keys_test.go, line 210 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17565#-KrCRBHysMpuibhe0na2:-KrCVFW7tQglDzEJqE0T:b-loldqa) ([raw file](https://github.com/cockroachdb/cockroach/blob/e74b74a0ba712d106321ae5cedca800821a67b09/pkg/keys/keys_test.go#L210)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

should these all be roachpb.Key instead of `[]byte`? ditto Reverse
</blockquote></details>

This actually resulted in a whole bunch of additional casting as some of the keys we're specifying have the wrong type and `MetaScanBounds` takes a `roachpb.RKey`. I'm going to leave as-is.

---

*[pkg/keys/spans.go, line 23 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17565#-KrCRQtRQsYwKWMvm2Ye:-KrCVUDK-LeNgh-QrQGm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e74b74a0ba712d106321ae5cedca800821a67b09/pkg/keys/spans.go#L23)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

was there supposed to be something after this colon?
</blockquote></details>

Done.

---

*[pkg/kv/split_test.go, line 108 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17565#-KrCRU08y3Bg-FSB5mKP:-KrCVVv2tbWMUQe3DoGy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e74b74a0ba712d106321ae5cedca800821a67b09/pkg/kv/split_test.go#L108)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

can you reference an issue? (throughout)
</blockquote></details>

Done.

---

*[pkg/storage/replica_command.go, line 1353 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17565#-KrCRyhJrDcg55VRJJnN:-KrCVE8fjitEzytaBbFe:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e74b74a0ba712d106321ae5cedca800821a67b09/pkg/storage/replica_command.go#L1353)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

nit: `'%[1]s' [%[1]x]` so you pass args.Key just once.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17565)*

**Comment by tamird** [comment]
Reviewed 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17565)*

**Comment by nvanbenschoten** [comment]
:lgtm:

---

Reviewed 2 of 7 files at r2, 3 of 4 files at r3.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[pkg/storage/replica_command.go, line 1334 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17565#-KrCaOovY54pYt-d_u2L:-KrCaOovY54pYt-d_u2M:bc61q2c) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ed29f32805fe34f3feb5a80675ae5e2ee1ba286/pkg/storage/replica_command.go#L1334)):*
> ```Go
> 		// No matching results were returned from the scan. This can happen when
> 		// meta2 ranges split (so for now such splitting is disabled). Consider the
> 		// scenario:
> ```

For the sake of anyone like me, could you mention that the meta copy of range descriptors is stored under the end key?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17565#-:-KrC_SdDp9S5XL_danw7:bnfp4nl)*

**Comment by petermattis** [comment]
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replica_command.go, line 1334 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17565#-KrCaOovY54pYt-d_u2L:-KrCeuK1qoOLgZv33EJz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ed29f32805fe34f3feb5a80675ae5e2ee1ba286/pkg/storage/replica_command.go#L1334)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

For the sake of anyone like me, could you mention that the meta copy of range descriptors is stored under the end key?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17565)*

File: pkg/storage/replica_command.go [hunk 1]
```diff
@@ -1304,12 +1329,30 @@ func evalRangeLookup(
...
+		//     f -> [e, f)
+		//     g -> [f, g)
+		//
+		// Now consider looking up the range containing key `d`. The DistSender
+		// routing logic would send the RangeLookup request to range 1 since `d`
+		// lies within the bounds of that range. But notice that the range
+		// descriptor containing `d` lies in range 2. Boom! A real fix will involve
+		// additional logic in the RangeDescriptorCache range lookup state machine.
+		//
+		// See #16266.
```
Comment (tschottdorf) [inline pkg/storage/replica_command.go]: or the much older https://github.com/cockroachdb/cockroach/issues/2266, but doesn't matter.

Comment (petermattis) [inline pkg/storage/replica_command.go]: I was planning to close #2266 as the primary issue there is being fixed (we won't access keys during `RangeLookup` that are outside are range) and leave #16266 for properly handling meta2 splits. Agreed that this doesn't really matter.

**Comment by bdarnell** [comment]
:lgtm:

---

Reviewed 2 of 7 files at r2, 3 of 4 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17565#-:-KrN93JRxyEMg3dXG41r:bnfp4nl)*

**Pull request merged**